crates/swc_ecma_utils/src/function.rs

```diff
@@ -0,0 +1,123 @@
+use swc_atoms::js_word;
+use swc_common::{util::take::Take, DUMMY_SP};
+use swc_ecma_ast::*;
+use swc_ecma_visit::{noop_visit_mut_type, VisitMut, VisitMutWith};
+
+#[derive(Default)]
+pub struct WrapperState {
+    this: Option<Ident>,
+    args: Option<Ident>,
+}
+
+impl WrapperState {
+    pub fn new(this: Option<Ident>, args: Option<Ident>) -> Self {
+        Self { this, args }
+    }
+    pub fn to_decl(self) -> Vec<VarDeclarator> {
+        let Self { this, args } = self;
+
+        let mut decls = Vec::with_capacity(2);
+        if let Some(this_id) = this {
+            decls.push(VarDeclarator {
+                span: DUMMY_SP,
+                name: Pat::Ident(this_id.clone().into()),
+                init: Some(Box::new(Expr::This(ThisExpr { span: DUMMY_SP }))),
+                definite: false,
+            });
+        }
+        if let Some(id) = args {
+            decls.push(VarDeclarator {
+                span: DUMMY_SP,
+                name: Pat::Ident(id.into()),
+                init: Some(Box::new(Expr::Ident(Ident::new(
+                    js_word!("arguments"),
+                    DUMMY_SP,
+                )))),
+                definite: false,
+            });
+        }
+        decls
+    }
+    pub fn to_stmt(self) -> Option<Stmt> {
+        let decls = self.to_decl();
+
+        if decls.is_empty() {
+            None
+        } else {
+            Some(Stmt::Decl(Decl::Var(VarDecl {
+                span: DUMMY_SP,
+                kind: VarDeclKind::Var,
+                declare: false,
+                decls,
+            })))
+        }
+    }
+}
+
+impl Take for WrapperState {
+    fn dummy() -> Self {
+        Self::default()
+    }
+}
+
+pub struct FunctionWrapper<'a> {
+    state: &'a mut WrapperState,
+}
+
+impl<'a> FunctionWrapper<'a> {
+    pub fn new(state: &'a mut WrapperState) -> Self {
+        FunctionWrapper { state }
+    }
+}
+
+impl<'a> VisitMut for FunctionWrapper<'a> {
+    noop_visit_mut_type!();
+
+    fn visit_mut_expr(&mut self, e: &mut Expr) {
+        match e {
+            Expr::Ident(id) if id.sym == js_word!("arguments") => {
+                let arguments = self.state.args.get_or_insert(private_ident!("_arguments"));
```

Comment:
This should be `get_or_insert_with(|| private_ident!("_arguments"))`. Otherwise, global lock will be acquired every time.

Reply:
This is new to me